File: app/src/main/java/com/aegamesi/steamtrade/steam/SteamUtil.java
package com.aegamesi.steamtrade.steam;

import android.util.Log;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;

public class SteamUtil {
	private final static char[] hexArray = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
	private final static HashMap<String, String> bbCodeMap = new HashMap<>();
	public static String webApiKey = null; // kept in secret.xml

	static {
		bbCodeMap.put("(?i)\\[b\\](.+?)\\[/b\\]", "<b>$1</b>");
		bbCodeMap.put("(?i)\\[i\\](.+?)\\[/i\\]", "<i>$1</i>");
		bbCodeMap.put("(?i)\\[u\\](.+?)\\[/u\\]", "<u>$1</u>");
		bbCodeMap.put("(?i)\\[h1\\](.+?)\\[/h1\\]", "<h5>$1</h5>");
		bbCodeMap.put("(?i)\\[spoiler\\](.+?)\\[/spoiler\\]", "[SPOILER: $1]");
		bbCodeMap.put("(?i)\\[strike\\](.+?)\\[/strike\\]", "<strike>$1</strike>");
		bbCodeMap.put("(?i)\\[url\\](.+?)\\[/url\\]", "<a href=\"$1\">$1</a>");
		//noparse still parses inner tags.
		bbCodeMap.put("(?i)\\[noparse\\](.+?)\\[/noparse\\]", "$1");
		//Purposely expose URL(Security), but have URL-Summary in brackets.
		bbCodeMap.put("(?i)\\[url=(.+?)\\](.+?)\\[/url\\]", "[$2] $1");
	}

	static byte[] calculateSHA1(byte[] data) {
		try {
			MessageDigest md = MessageDigest.getInstance("SHA1");
			return md.digest(data);
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		}
		return null;
	}

	public static String bytesToHex(byte[] bytes) {
		if (bytes == null)
			return "0000000000000000000000000000000000000000";
		char[] hexChars = new char[bytes.length * 2];
		int v;
		for (int j = 0; j < bytes.length; j++) {
			v = bytes[j] & 0xFF;
			hexChars[j * 2] = hexArray[v >>> 4];
			hexChars[j * 2 + 1] = hexArray[v & 0x0F];
		}
		return new String(hexChars);
	}

	public static String parseBBCode(String source) {
		source = parseEmoticons(source);

		for (Map.Entry entry : bbCodeMap.entrySet())
			source = source.replaceAll(entry.getKey().toString(), entry.getValue().toString());

		Log.d("BB", source);
		return source;
	}

	public static String parseEmoticons(String source) {
		return source.replaceAll("\u02D0([a-zA-Z0-9_]+)\u02D0", "<img src=\"http://steamcommunity-a.akamaihd.net/economy/emoticon/$1\">").replaceAll("(\r\n|\r|\n|\n\r)", "<br/>");
	}

}

File: app/src/main/java/com/aegamesi/steamtrade/steam/SteamUtil.kt
package com.aegamesi.steamtrade.steam

import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*

object SteamUtil {
    private val hexArray =
            charArrayOf(
                    '0', '1', '2', '3', '4',
                    '5', '6', '7', '8', '9',
                    'A', 'B', 'C', 'D', 'E', 'F')
    private val bbCodeMap = HashMap<String, String>()
    var webApiKey: String? = null // kept in secret.xml

    init {
        bbCodeMap["(?i)\\[b\\](.+?)\\[/b\\]"] = "<b>$1</b>"
        bbCodeMap["(?i)\\[i\\](.+?)\\[/i\\]"] = "<i>$1</i>"
        bbCodeMap["(?i)\\[u\\](.+?)\\[/u\\]"] = "<u>$1</u>"
        bbCodeMap["(?i)\\[h1\\](.+?)\\[/h1\\]"] = "<h5>$1</h5>"
        bbCodeMap["(?i)\\[spoiler\\](.+?)\\[/spoiler\\]"] = "[SPOILER: $1]"
        bbCodeMap["(?i)\\[strike\\](.+?)\\[/strike\\]"] = "<strike>$1</strike>"
        bbCodeMap["(?i)\\[url\\](.+?)\\[/url\\]"] = "<a href=\"$1\">$1</a>"
        //noparse still parses inner tags.
        bbCodeMap["(?i)\\[noparse\\](.+?)\\[/noparse\\]"] = "$1"
        //Purposely expose URL(Security), but have URL-Summary in brackets.
        bbCodeMap["(?i)\\[url=(.+?)\\](.+?)\\[/url\\]"] = "[$2] $1"
    }

    internal fun calculateSHA1(data: ByteArray): ByteArray? {
        try {
            val md = MessageDigest.getInstance("SHA1")
            return md.digest(data)
        } catch (e: NoSuchAlgorithmException) {
            e.printStackTrace()
        }

        return null
    }

    fun bytesToHex(bytes: ByteArray?): String {
        if (bytes == null)
            return "0000000000000000000000000000000000000000"
        val hexChars = CharArray(bytes.size * 2)
        var v: Int
        for (j in bytes.indices) {
            v = bytes[j].toInt() and 0xFF
            hexChars[j * 2] = hexArray[v.ushr(4)]
            hexChars[j * 2 + 1] = hexArray[v and 0x0F]
        }
        return String(hexChars)
    }

    fun parseBBCode(src: String): String {
        var source = parseEmoticons(src)

        for ((key, value) in bbCodeMap)
            source = source.replace(key.toRegex(), value)

        return source
    }

    fun parseEmoticons(source: String): String {
        return source.replace(
                "\u02D0([a-zA-Z0-9_]+)\u02D0".toRegex(),
                "<img src=\"http://steamcommunity-a.akamaihd.net/economy/emoticon/$1\">")
                .replace("(\r\n|\r|\n|\n\r)"
                .toRegex(), "<br/>")
    }

    fun getAvatar(source: ByteArray?): String {
        val baseURL = "http://media.steampowered.com/steamcommunity/public/images/avatars/"
        val nullImage = "fef49e7fa7e1997310d705b2a6158ff8dc1cdfeb"
        val allZERO = "0000000000000000000000000000000000000000"

        val imgHash = bytesToHex(source).toLowerCase(Locale.US)

        return if (imgHash != allZERO && imgHash.length == 40) {
            baseURL + imgHash.substring(0, 2) + "/" + imgHash + "_full.jpg"
        }
        else{
            baseURL + imgHash.substring(0, 2) + "/" + nullImage + "_full.jpg"
        }
    }

}
